File: examples/Actor.Next/06-Approvals/ApprovalCapabilities.cs
using System.Text.Json;
using Approvals.Next.Example06.Effects;
using Approvals.Next.Example06.Guards;
using Dapr.Actors.Next.Abstractions.Filters;
using Dapr.Actors.Next.Interpreted;
using Dapr.Workflow;
using Microsoft.Extensions.Logging.Abstractions;

namespace Dapr.Actors.Next.Examples.Approvals;

/// <summary>
/// The vetted guards and effects an interpreted approval machine may reference by name. A definition
/// can only compose from these, so a typo or an unsupported action fails verification at onboarding
/// rather than on a live document.
/// </summary>
public sealed class ApprovalCapabilityRegistry : ICapabilityRegistry
{
    /// <summary>The auto-approval limit; amounts at or below it skip manager escalation.</summary>
    public const decimal AutoApprovalLimit = 1_000m;

    private readonly Dictionary<string, IActorEffect> _effects = new(StringComparer.Ordinal);
    private readonly Dictionary<string, IActorGuard> _guards = new(StringComparer.Ordinal);

    public ApprovalCapabilityRegistry(IDaprWorkflowClient workflowClient, ILoggerFactory? loggerFactory = null)
    {
        ArgumentNullException.ThrowIfNull(workflowClient);
        var logger = (loggerFactory ?? NullLoggerFactory.Instance).CreateLogger<ApprovalCapabilityRegistry>();

        _effects[ApprovalDefinitions.RecordSubmissionEffect] = new RecordSubmissionEffect();
        _effects[ApprovalDefinitions.RecordDecisionEffect] = new RecordDecisionEffect();
        _effects[ApprovalDefinitions.NotifyManagerEffect] = new LogEffect(logger, "escalated to a manager");
        _effects[ApprovalDefinitions.RecordSettlementFailureEffect] = new RecordSettlementFailureEffect();
        _effects[ApprovalDefinitions.StartSettlementEffect] = new StartSettlementEffect(workflowClient, logger);

        _guards[ApprovalDefinitions.WithinApprovalLimitGuard] = new WithinApprovalLimitGuard();
    }

    public bool TryGetEffect(string name, out IActorEffect effect) => _effects.TryGetValue(name, out effect!);

    public bool TryGetGuard(string name, out IActorGuard guard) => _guards.TryGetValue(name, out guard!);

    internal static DynamicStateBag State(ActorCapabilityContext context) =>
        (DynamicStateBag)context.Arguments["state"]!;

    internal static JsonElement Payload(ActorCapabilityContext context) =>
        (JsonElement)context.Arguments["payload"]!;
}

File: examples/Actor.Next.FSharp/06-Approvals/Approvals.Next.FSharp.Example06.Actors/ApprovalCapabilities.fs
namespace Dapr.Actors.Next.Examples.Approvals

open System
open System.Collections.Generic
open Microsoft.Extensions.Logging
open Microsoft.Extensions.Logging.Abstractions
open Dapr.Actors.Next.Abstractions.Filters
open Dapr.Workflow

type ApprovalCapabilityRegistry(workflowClient: IDaprWorkflowClient, ?loggerFactory: ILoggerFactory) =
    let loggerFactory = defaultArg loggerFactory (NullLoggerFactory.Instance :> ILoggerFactory)
    let logger = loggerFactory.CreateLogger<ApprovalCapabilityRegistry>()

    let effects = Dictionary<string, IActorEffect>(StringComparer.Ordinal)
    let guards = Dictionary<string, IActorGuard>(StringComparer.Ordinal)

    do
        effects.[ApprovalDefinitions.RecordSubmissionEffect] <- RecordSubmissionEffect() :> IActorEffect
        effects.[ApprovalDefinitions.RecordDecisionEffect] <- RecordDecisionEffect() :> IActorEffect
        effects.[ApprovalDefinitions.NotifyManagerEffect] <- LogEffect(logger, "escalated to a manager") :> IActorEffect
        effects.[ApprovalDefinitions.RecordSettlementFailureEffect] <- RecordSettlementFailureEffect() :> IActorEffect
        effects.[ApprovalDefinitions.StartSettlementEffect] <- StartSettlementEffect(workflowClient, logger) :> IActorEffect

        guards.[ApprovalDefinitions.WithinApprovalLimitGuard] <- WithinApprovalLimitGuard() :> IActorGuard

    interface ICapabilityRegistry with
        member _.TryGetEffect(name: string, effect: byref<IActorEffect>) : bool =
            match effects.TryGetValue(name) with
            | true, e -> effect <- e; true
            | false, _ -> effect <- Unchecked.defaultof<_>; false

        member _.TryGetGuard(name: string, guard: byref<IActorGuard>) : bool =
            match guards.TryGetValue(name) with
            | true, g -> guard <- g; true
            | false, _ -> guard <- Unchecked.defaultof<_>; false
